explorer/interpreter/type_checker.cpp

```diff
@@ -5291,57 +5291,54 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
     class_decl->set_static_type(&self->static_type());
   }
 
-  // Find base class declaration, if any. Verify that is before any data member
-  // declarations, and there is at most one.
+  // Find base class declaration, if any. Right now, verify that it is first in
+  // the class. This avoids the problem identified in
+  // https://github.com/carbon-language/carbon-lang/issues/2994 where the base
+  // class expression could reference an earlier declaration in the class that
+  // hasn't been typechecked yet and therefore doesn't have its `static_type`
+  // set.
+
+  // TODO: Verify just that is before any data member declarations, and there is
+  // at most one, and delay remaining work (type checking, base class
+  // evaluation, etc.) until the `extend base` declaration is processed in
+  // order.
   std::optional<Nonnull<const NominalClassType*>> base_class;
-  bool after_data_member = false;
+  bool first = true;
```

Comment:
Would it work to process `class_decl->members()[0]` before the loop, and then loop over the remaining members to check that they're not `ExtendBaseDeclaration`s? That seems like it would be clearer to me, because the logic is expressed more in the static structure of the code rather than in a dynamic state variable. It might also be a little safer, because there's no risk of forgetting to update the state variable.

Reply:
It isn't clearly better to me, but I made the change.